File: BTRightSideView_BFS.py
# TC: O(N) since we are processing all the nodes
# SC: O(N) where N is the size of the queue at any instance

class Solution:
    def rightSideView(self, root: TreeNode) -> List[int]:
        self.final = []
        
        if not root:
            return self.final
            
        queue = collections.deque()
        queue.append(root)
        while queue:
            size = len(queue)
            for i in range(size):
                curr = queue.popleft()
                if i == size - 1:
                    self.final.append(curr.val)
                
                if curr.left:
                    queue.append(curr.left)
                if curr.right:
                    queue.append(curr.right)
        
        return self.final

File: BTRightSideViewBFS.java
//Time Complexity - O(n)
//Space complexity - O(h) - height of the tree

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<Integer> result;
    public List<Integer> rightSideView(TreeNode root) {
      result = new ArrayList<>();
      //BFS solution
      Queue <TreeNode> q = new LinkedList<>();
      q.add(root);
      
      while(!q.isEmpty()) {
        int size = q.size();
        //add childeren to the queue
        for(int i=0; i< size;i++) {
          TreeNode curr = q.poll();
          System.out.println("size: " + size + " i: " + i);
          //if its last node then add it to the result
          if(i == size-1) {
            result.add(curr.val);
          }
          //add left and right children in queue
          if(curr.left!=null) q.add(curr.left);
          if(curr.right!=null) q.add(curr.right);
        }
      }
        
      return result;
    }
}